src/Service/CleanupSnapshotService.php

```diff
@@ -35,16 +34,12 @@ public function cleanupBySite(SiteInterface $site, int $keepSnapshots): void
     {
         $pages = $this->pageManager->findBy(['site' => $site->getId()]);
 
-        if ($this->snapshotManager instanceof BaseEntityManager) {
-            $this->snapshotManager->getEntityManager()->beginTransaction();
-        }
+        $this->snapshotManager->beginTransaction();
```

Comment:
if ($this->snapshotManager instanceof TransactionalManagerInterface) {
    $this->snapshotManager->beginTransaction();
}

Reply:
if I do this it won't work, this method will not do anything! IMO it doesn't make sense to add this "if"

What do you think to create `SnapshotEntityManagerInterface` and extends it from `SnapshotManagerInterface` and `TransactionalManagerInterface`?

Reply:
I agree with @VincentLanglet , this interface was introduced precisely for that reason, we will only do transactions if the manager supports them. You do not want the model folder to force the use of the interface, because someone could implement MongoDB (without transactions) and it will be forced to add those methods, even if he knows they make no sense in its implementation.

Reply:
Ok I got your point, But this "concrete" class was build to work only with ORM! if the user wants to make this work for ODM, then he should create a new class implementing `CleanupSnapshotBySiteInterface`.

if you look into the implementation we gonna have more problems if the user try to use ODM here like

```php
$pages = $this->pageManager->findBy(['site' => $site->getId()]);
```

it will broken too, this class definitely doesn't expect to receive an `ODM`


if `SnapshotManagerInterface` is used for `ORM` and `ODM`, it's fine, let's create an abstraction of this class like `SnapshotEntityManagerInterface` and require this into the constructor!

Reply:
> Ok I got your point, But this "concrete" class was build to work only with ORM! if the user wants to make this work for ODM, then he should create a new class implementing `CleanupSnapshotBySiteInterface`.
> 
> if you look into the implementation we gonna have more problems if the user try to use ODM here like
> 
> ```
> $pages = $this->pageManager->findBy(['site' => $site->getId()]);
> ```
> 
> it will broken too, this class definitely doesn't expect to receive an `ODM`
> 
> if `SnapshotManagerInterface` is used for `ORM` and `ODM`, it's fine, let's create an abstraction of this class like `SnapshotEntityManagerInterface` and require this into the constructor!

maybe we could add such examples into the doc

Reply:
> Ok I got your point, But this "concrete" class was build to work only with ORM! if the user wants to make this work for ODM, then he should create a new class implementing `CleanupSnapshotBySiteInterface`.

The class/interface was not build to only work with ORM. That's all the purpose of using abstraction from sonata-doctrine-extensions.

> if you look into the implementation we gonna have more problems if the user try to use ODM here like
> 
> ```
> $pages = $this->pageManager->findBy(['site' => $site->getId()]);
> ```
> 
> it will broken too, this class definitely doesn't expect to receive an `ODM`

https://github.com/doctrine/mongodb-odm/blob/2.4.x/lib/Doctrine/ODM/MongoDB/Repository/DocumentRepository.php#L172

I don't know MongoDB a lot but findBy exists in ODM

> if `SnapshotManagerInterface` is used for `ORM` and `ODM`, it's fine, let's create an abstraction of this class like `SnapshotEntityManagerInterface` and require this into the constructor!

I prefer to let possible the fact that one day ODM will be supported, rather by restricting to ORM in the constructor.

Reply:
> I prefer to let possible the fact that one day ODM will be supported, rather by restricting to ORM in the constructor.

But the `CleanupSnapshotBySiteInterface` can support ODM, he just need to decorate this class [src/Service/CleanupSnapshotService.php](https://github.com/sonata-project/SonataPageBundle/pull/1550/files/69e68b0627e541e78ae32779b3edb2ad04e0516a#diff-112198c8b3816a6af7235c0c526ba39544d633700e4538d9ab1fc7e7c770d8f6)!

But let's imagine that someone going to implement something to support `clean snapshot` for ODM! then he needs to make conditions to check if it's ORM or ODM, don't you think that should be better he Create a new class and just implement the `CleanupSnapshotBySiteInterface`?

Reply:
and what's happen if I want to use other thing different to save my snapshots? like redis, then should I check if it's `ORM`, `ODM` or `Redis`?

Reply:
> But the `CleanupSnapshotBySiteInterface` can support ODM, he just need to decorate this class [src/Service/CleanupSnapshotService.php](https://github.com/sonata-project/SonataPageBundle/pull/1550/files/69e68b0627e541e78ae32779b3edb2ad04e0516a#diff-112198c8b3816a6af7235c0c526ba39544d633700e4538d9ab1fc7e7c770d8f6)!

With the check
```
if ($this->snapshotManager instanceof TransactionalManagerInterface)
```
he won't have to decorate at all.

> But let's imagine that someone going to implement something to support `clean snapshot` for ODM! then he needs to make conditions to check if it's ORM or ODM, don't you think that should be better he Create a new class and just implement the `CleanupSnapshotBySiteInterface`?

I don't see why he need to check ORM or ODM. The goal is that this class support both.

Reply:
In theory, to implement another persistency strategy, you just need to replicate the "Entity" folder (Which is specific for doctrine ORM) and name it whatever you want (Redis, Document for ODM), take a look at other bundles like SonataMediaBundle, SonataClassificationBundle and SonataUserBundle.

Reply:
See https://github.com/sonata-project/SonataMediaBundle/tree/4.x/src/Document as an exemple.

There is a GalleryManager which implements the GalleryManagerInterface (defined in the Model folder).
If the GalleryManagerINterface would have extends the TransactionalManagerInterface it would have block the ODM support.

Reply:
> In theory, to implement another persistency strategy, you just need to replicate the "Entity" folder (Which is specific for doctrine ORM) and name it whatever you want (Redis, Document for ODM), take a look at other bundles like SonataMediaBundle, SonataClassificationBundle and SonataUserBundle.

- Could you provide one example?

- But look I do not want to change all my persist, just the snapshot stuffs because I could use "Doctrine ORM" to save my data for others entities, But the snapshot json, I want to save in redis!

is it possible to do this with doctrine?

Reply:
> There is a GalleryManager which implements the GalleryManagerInterface (defined in the Model folder).
> If the GalleryManagerINterface would have extends the TransactionalManagerInterface it would have block the ODM support.

I checked I got the idea, I know that `GalleryManagerINterface` should works for both, like `SnapshotManagerInterface`, But I still doesn't agree with this 😄 for me no make sense depends of an Interface that need to add "ifs" to check which code should be executed! IMO we should depends the interface that offers that methods that are necessary for my code works! IMO it's a good start to let the class do more that would suppose to do!

Reply:
But well here we have 2 votes x 1 vote 😄 

I guess it won't implement more then `ORD` and `ODM` and in case I want to do something customised like check in different place I can still make my own implementation!

Reply:
Ok But as ATM it just work for `ORM` can I check throw an exeption in case he is not using `ORM`? when someone implements the ODM, then we can remove the `Exception`!

if yes, which exception should I return?

Reply:
No, we need to maintain the if with the check for instanceof, that the most compatible solution.

Reply:
I don't see why an exception should be thrown 

>  But I still doesn't agree with this 😄 for me no make sense depends of an Interface that need to add "ifs" to check which code should be executed!  IMO we should depends the interface that offers that methods that are necessary for my code works!

If it implements Transactional, you run the code inside a transaction
If it don't, you still run the code, but without beginTransaction and commit. The code will still work and save the things.

So the Transactional is not mandatory.